File: Interval Minimum Number.cpp
/**
 * Definition of Interval:
 * classs Interval {
 *     int start, end;
 *     Interval(int start, int end) {
 *         this->start = start;
 *         this->end = end;
 *     }
 */
class SegmentTreeMiniNode {
public:
    int start, end, mini;
    SegmentTreeMiniNode *left, *right;
    SegmentTreeMiniNode(int start, int end)
    {
        this->start = start;
        this->end = end;
        this->mini = INT_MIN;
        this->left = this->right = nullptr;
    }
};
class Solution { 
private:
    SegmentTreeMiniNode *root_;
public:
    /**
     *@param A, queries: Given an integer array and an query list
     *@return: The result list
     */
    vector<int> intervalMinNumber(vector<int> &A, vector<Interval> &queries) {
        // write your code here
        root_ = build(0, A.size() - 1, A);
        vector<int> ret;
        for (const auto &interval : queries)
        {
            ret.emplace_back(query(interval.start, interval.end, root_));
        }
        return ret;
    }
    int query(int start, int end, SegmentTreeMiniNode *root)
    {
        if (!root || root->start > end || root->end < start)
        {
            return INT_MAX;
        }
        else
        {
            if (start <= root->start && end >= root->end)
            {
                return root->mini;
            }
            else
            {
                return min(query(start, end, root->left), query(start, end, root->right));
            }
        }
    }
    SegmentTreeMiniNode *build(int start, int end, vector<int> &A)
    {
        if (start > end)
        {
            return nullptr;
        }
        else
        {
            SegmentTreeMiniNode *root = new SegmentTreeMiniNode(start, end);
            if (start == end)
            {
                root->mini = A[start];
                return root;
            }
            else
            {
                int mid = start + (end - start) / 2;
                root->left = build(start, mid, A);
                root->right = build(mid + 1, end, A);
                int left_mini = root->left ? root->left->mini : INT_MAX;
                int right_mini = root->right ? root->right->mini : INT_MAX;
                root->mini = min(left_mini, right_mini);
                return root;
            }
        }
    }
};

File: Interval Minimum Number.py
"""
Definition of Interval.
class Interval(object):
    def __init__(self, start, end):
        self.start = start
        self.end = end
"""

class SegNode:
    def __init__(self, start, end):
        self.start, self.end, self.min = start, end, None
        self.left, self.right = None, None
        
class Solution:	
    """
    @param A, queries: Given an integer array and an Interval list
                       The ith query is [queries[i-1].start, queries[i-1].end]
    @return: The result list
    """
    def intervalMinNumber(self, A, queries):
        # write your code here
        def doBuild(left, right):
            node = SegNode(left, right)
            if left == right:
                node.min = A[left]
                return node
            else:
                mid = left + (right - left) / 2
                node.left = doBuild(left, mid)
                node.right = doBuild(mid + 1, right)
                node.min = min(node.left.min, node.right.min)
                return node
        def doQuery(node, start, end):
            if node.start > end or node.end < start:
                return float('inf')
            else:
                if node.start >= start and node.end <= end:
                    return node.min
                else:
                    return min(doQuery(node.left, start, end), doQuery(node.right, start, end))
        
        root = doBuild(0, len(A) - 1)
        ret = []
        for q in queries:
            ret.append(doQuery(root, q.start, q.end))
        return ret
            